add db staleness check

jonasagx opened the pull request:
* Add staleness check to vuln db. Execution fails if db is 5 or more days old.  Configurable via `grype.yaml` and enabled by default. 
* Warn user that db is stale even when staleness check is disabled.
* Add version banner comparing application version with new version available illustrating  to users how far behind their local grype is.

FIx: #240 

Note on staleness field type: it uses go's time duration notation. Anchore provides daily updates. However users might use grype with a different database and different release schedules. Go's duration notation has flexible granularity to different use cases. It also helps maintaining grype's config file consistent over time.

Signed-off-by: Jonas Xavier <jonasx@anchore.com>

luhring commented on grype/db/curator.go:
```diff
@@ -267,6 +278,24 @@ func (c *Curator) download(listing *ListingEntry, downloadProgress *progress.Man
+		return nil
+	}
+
+	age := time.Since(m.Built).Round(time.Hour)
```
What is the motivation for rounding here?

jonasagx replied:
I added a code comment explaining it; 

tl;dr time comparison has fractional seconds of precision which makes the error message noisy. Rounding it to hour provides enough context for action.

jonasagx replied:
Rounding still happens but through a time formatting library capable of printing longer durations in a more friendly way, like `22 days` instead of `528h`

luhring commented on grype/db/curator.go:
```diff
@@ -267,6 +278,24 @@ func (c *Curator) download(listing *ListingEntry, downloadProgress *progress.Man
+	}
+
+	if age > DefaultStalenessThreshold {
+		log.Warnf("your vuln db was updated %s ago and it might generate wrong results. To update run: grype db update", age)
```
This message could be improved to be more clear and idiomatic to our log messages:

1. The problem isn't when the database was updated, it's when it was built.
2. I'd say "inaccurate" instead of something as simplified as "wrong".
3. We typically don't say "your vuln db". Instead we tend to say "the vulnerability database".

jonasagx replied:
Thank you for the language clarifications, I applied the suggestions.

luhring commented on cmd/root.go:
```diff
@@ -270,7 +270,7 @@ func startWorker(userInput string, failOnSeverity *vulnerability.Severity) <-cha
 			}
 			if isAvailable {
-				log.Infof("New version of %s is available: %s", internal.ApplicationName, newVersion)
+				log.Infof("You're currently running %s version %s and a new version is available: %s", internal.ApplicationName, version.FromBuild().Version, newVersion)
```
(Same comment as the other instance of this message)

jonasagx replied:
Yes, it reads better with the above suggestion, ty.

luhring commented on README.md:
```diff
@@ -399,6 +399,10 @@ By default, the database is cached on the local filesystem in the directory `$XD
 
+#### Data staleness
+
+Grype needs up-to-date vulnerability information to provide good matches. By default, it will fail execution if the local database was not updated in the last 5 days. The data staleness check is configurable via the field `staleness-threshold`, under `db`. It uses golang's time duration syntax. Set it to 0 to disable staleness check.
```
I'd say "accurate matches" instead of "good matches".

luhring commented on README.md:
```diff
@@ -399,6 +399,10 @@ By default, the database is cached on the local filesystem in the directory `$XD
 
+#### Data staleness
+
+Grype needs up-to-date vulnerability information to provide good matches. By default, it will fail execution if the local database was not updated in the last 5 days. The data staleness check is configurable via the field `staleness-threshold`, under `db`. It uses golang's time duration syntax. Set it to 0 to disable staleness check.
```
>It uses golang's time duration syntax

It'd be helpful to link to where this is described, in case the user isn't familiar with this syntax.

jonasagx replied:
Agreed, I added the link

luhring commented on grype/db/curator.go:
```diff
@@ -28,11 +30,18 @@ const (
 	ListingURL          string
 	CACert              string
 	ValidateByHashOnGet bool
+	StalenessThreshold  time.Duration
```
It's not immediately obvious what "staleness threshold" means. We should be exact here. We want this value to represent the max time allowed since the database was built.

luhring commented on internal/config/database.go:
```diff
@@ -11,11 +12,12 @@ import (
+	CACert                string        `yaml:"ca-cert" json:"ca-cert" mapstructure:"ca-cert"`
+	AutoUpdate            bool          `yaml:"auto-update" json:"auto-update" mapstructure:"auto-update"`
+	ValidateByHashOnStart bool          `yaml:"validate-by-hash-on-start" json:"validate-by-hash-on-start" mapstructure:"validate-by-hash-on-start"`
+	StalenessThreshold    time.Duration `yaml:"staleness-threshold" json:"staleness-threshold" mapstructure:"staleness-threshold"`
```
It's not immediately obvious what "staleness threshold" means. We should be exact here. We want this value to represent the max time allowed since the database was built.

jonasagx replied:
I agree, I renamed it to `MaxAllowedDBAge`, which I think is more explicit and direct. Please let me know of any possible redundancy.

luhring reviewed (changes requested):
Thanks for getting this started! I left a few notes on clarity, logic, and wording. Let me know if you have any questions.

jonasagx commented on grype/db/curator.go:
```diff
@@ -62,6 +72,7 @@ func NewCurator(cfg Config) (Curator, error) {
 		dbPath:              path.Join(dbDir, FileName),
 		listingURL:          cfg.ListingURL,
 		validateByHashOnGet: cfg.ValidateByHashOnGet,
+		dataStalenessLimit:  cfg.StalenessThreshold,
```
Agreed, I renamed the fields accordingly.

jonasagx commented on grype/db/curator_test.go:
```diff
@@ -301,3 +301,60 @@ func TestCuratorDBPathHasSchemaVersion(t *testing.T) {
+	return func(t assert.TestingT, err error, i ...interface{}) bool {
+		return assert.ErrorContains(t, err, expected)
+	}
+}
```
That makes sense. After some code and testing refactoring this helper became unnecessary.

jonasagx commented on grype/db/curator_test.go:
```diff
@@ -301,3 +301,60 @@ func TestCuratorDBPathHasSchemaVersion(t *testing.T) {
+			name: "stale-data",
+			fields: fields{
+				staleLimist: time.Hour,
+				md:          &Metadata{Built: time.Now().Add(-4 * time.Hour)},
```
In go `time.Sub` returns the duration between two times, while `time.Add` adds a duration to a time. Here `time.Time` is made from a given duration and why `time.Add` is necessary.

luhring replied:
Good call!

jonasagx commented on internal/ui/etui_event_handlers.go:
```diff
@@ -28,7 +29,7 @@ func handleAppUpdateAvailable(_ context.Context, fr *frame.Frame, event partybus
 	}
 
-	message := color.Magenta.Sprintf("New version of %s is available: %s", internal.ApplicationName, newVersion)
+	message := color.Magenta.Sprintf("You're currently running %s version %s and a new version is available: %s", internal.ApplicationName, version.FromBuild().Version, newVersion)
```
Yes, it reads better with that suggestion, ty.

jonasagx commented on grype/db/curator.go:
```diff
@@ -267,6 +278,24 @@ func (c *Curator) download(listing *ListingEntry, downloadProgress *progress.Man
 	return tempDir, nil
 }
 
+func (c *Curator) validateStaleness(m *Metadata) error {
```
Sure, I added code comments to it.

jonasagx commented on grype/db/curator.go:
```diff
@@ -267,6 +278,24 @@ func (c *Curator) download(listing *ListingEntry, downloadProgress *progress.Man
+	age := time.Since(m.Built).Round(time.Hour)
+	// dataStalenessLimit == 0 means to not error on staleness
+	if c.dataStalenessLimit > 0 && age > c.dataStalenessLimit {
+		return fmt.Errorf("%w: last updated %s ago (> threshold %s)", ErrDataIsStale, age, c.dataStalenessLimit)
```
I dropped `ErrDataIsStale` (it was used only here anyways) in favor of creating it locally and added context information: `fmt.Errorf("the vulnerability database was built %s ago (> max allowed %s)", age, c.maxAllowedDBAge)`

wagoodman commented on README.md:
```diff
@@ -584,6 +588,11 @@ db:
+  # age being the time since it was built
+  # Default max age is 120h (or five days)
+  # Set it to 0 (zero) to disable check
+  max-allowed-db-age: "120h"
```
Can we additionally allow for `d` for days for config input? I realize that go `time.ParseDuration` doesn't handle this, but it's a little awkward to configure high values.

I've got similar comments around the output that we are indicating, for example:
```
$ GRYPE_DB_AUTO_UPDATE=false grype alpine:latest
1 error occurred:
	* failed to load vulnerability db: vulnerability database is invalid (run db update to correct): the vulnerability database was built 539h0m0s ago (> max allowed 120h0m0s)
```

Something like `539h0m0s` should get converted to `22 days` (or similar) since this message is intended for the user. Consider using something like https://github.com/dustin/go-humanize#times to help with showing human readable values.

jonasagx replied:
There is no general way to parse a duration like `1d` because days don't have consistent length. More on https://github.com/golang/go/issues/11473#issuecomment-116930849

I am in favor of human friendly outputs 👍🏼

jonasagx replied:
I've added https://github.com/hako/durafmt since its implementation doesn't assume local timezone (like https://github.com/dustin/go-humanize#times does). We have to compare duration to UTC; the db built value is in UTC.

wagoodman commented on README.md:
```diff
@@ -584,6 +588,11 @@ db:
+  # age being the time since it was built
+  # Default max age is 120h (or five days)
+  # Set it to 0 (zero) to disable check
+  max-allowed-db-age: "120h"
```
since this is already under the `DB` section, to prevent a stutter renaming to `max-allowed-age` I think makes sense -- what's your take?

jonasagx replied:
I changed it to `max-allowed-built-age`.

wagoodman commented:
Are we intending to block the `db import` path? That's the behavior I'm seeing:
```
$ grype db import ./ReallyOldDB.tar.gz
unable to import vulnerability database: the vulnerability database was built 539h0m0s ago (> max allowed 120h0m0s)
```

my opinion is that we should allow the user to import any DB they'd like (since the importing doesn't affect results) but when we run grype to get vulnerability results we should warn in that case (when the data is used).

jonasagx commented on README.md:
```diff
@@ -584,6 +588,11 @@ db:
   # same as GRYPE_DB_UPDATE_URL env var
   update-url: "https://toolbox-data.anchore.io/grype/databases/listing.json"
 
+  # Max allowed age for vulnerability database,
```
I added `validate-age` config field 👍🏼

luhring replied:
Should that new field get added to this config documentation here?

luhring replied:
Also, does this new config value _replace_ the "set to 0 to disable" mechanism?

wagoodman replied:
it probably should be mutually exclusive, good catch

jonasagx commented:
> I'm seeing two warnings instead of one when setting a large value and disabling the DB updating
I fixed that.

Other comment:
> my opinion is that we should allow the user to import any DB they'd like (since the importing doesn't affect results) but when we run grype to get vulnerability results we should warn in that case (when the data is used).
I agree with you and changed to only check staleness before a vuln scan.

luhring commented on cmd/root.go:
```diff
@@ -270,7 +270,7 @@ func startWorker(userInput string, failOnSeverity *vulnerability.Severity) <-cha
 			}
 			if isAvailable {
-				log.Infof("New version of %s is available: %s", internal.ApplicationName, newVersion)
+				log.Infof("new version of %s is available: %s (current version is %s)", internal.ApplicationName, newVersion, version.FromBuild().Version)
```
nit: I think "current version" is slightly ambiguous here. We've used "current version" in conversation before to refer to the _latest_ released version. I'd suggest something like `"<version> currently installed"`

jonasagx replied:
👍🏼  I changed it to `new version of %s is available: %s (currently running: %s)"` so the two versions are not next to each other helping visually.

luhring commented on grype/db/curator.go:
```diff
@@ -28,11 +30,18 @@ const (
+var (
+	ErrDataIsStale = errors.New("data is stale")
+
+	DefaultStalenessThreshold = time.Hour * 24 * 5
```
(@jonasagx looks like this comment might have been missed — this question is about exporting)

jonasagx replied:
@luhring `DefaultMaxAllowedBuiltAge` is exported for use in two packages: `grype/db/curator` and `internal/config/db`

jonasagx replied:
You're right, it _was_ used by both, but now only referenced in `internal/config`. I moved it there

luhring commented on internal/ui/etui_event_handlers.go:
```diff
@@ -28,7 +29,7 @@ func handleAppUpdateAvailable(_ context.Context, fr *frame.Frame, event partybus
 	}
 
-	message := color.Magenta.Sprintf("New version of %s is available: %s", internal.ApplicationName, newVersion)
+	message := color.Magenta.Sprintf("New version of %s is available: %s (current version is %s)", internal.ApplicationName, newVersion, version.FromBuild().Version)
```
(see comment in other instance of this message. Also, should we consider factoring out this message into a shared value?)

wagoodman replied:
my take, it's better as-is (not being a shared constant); the ETUI and logging are different-enough of concerns to not need a single source of truth for this message (and they have their own rules for how something is displayed). For instances, in logging the sentence should start with a lower case but in the ETUI it should be capitalized.

luhring commented on grype/db/curator.go:
```diff
@@ -28,11 +30,18 @@ const (
 )
 
+var (
+	ErrDataIsStale = errors.New("data is stale")
```
nit: I'd suggest a slight wording tweak -> `(max allowed age is %s)`

jonasagx replied:
👍🏼  I changed it to `(max allowed age is %s)`

wagoodman commented on grype/db/curator.go:
```diff
@@ -214,7 +226,12 @@ func (c *Curator) UpdateTo(listing *ListingEntry, downloadProgress, importProgre
+		return err
+	}
+
+	return c.validateStaleness(metadata)
```
it seems odd that `validateStaleness` is not part of the `validate` call -- if there is a specific reason it should probably be added as a comment here. Also if `validate` isn't covering all validations it should probably be renamed to the scope that it's responsible for, such as `validateIntegrity` or similar.

jonasagx replied:
Yes, I renamed it to `validateIntegrity`. That is the relevant check while importing a database for example.

luhring reviewed (approved):
Looks good, thanks @jonasagx!
